File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/inventory.ts
/**
 * Campaign inventory and character equipment assignment. Party stock is
 * tracked separately from per-character assignments; the summary reports
 * remaining healing-potion stock as party quantity minus assigned quantity.
 */

import { getDb } from '../db.ts';
import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

export function addInventoryItem(campaignId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const itemSlug = body.item_slug;
  if (typeof itemSlug !== 'string' || itemSlug.length === 0) {
    return { status: 400, body: { error: 'item_slug must be a non-empty string' } };
  }

  const quantity = body.quantity;
  if (!isValidIntInRange(quantity, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'quantity must be a positive integer' } };
  }

  const owner = body.owner;
  if (typeof owner !== 'string' || owner.length === 0) {
    return { status: 400, body: { error: 'owner must be a non-empty string' } };
  }

  db.prepare(
    `INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity)
     VALUES (?, ?, ?, ?)
     ON CONFLICT (campaign_id, item_slug, owner) DO UPDATE SET quantity = quantity + excluded.quantity`,
  ).run(campaignId, itemSlug, owner, quantity);

  return { status: 201, body: { item_slug: itemSlug, quantity, owner } };
}

export function assignEquipment(campaignId: string, characterId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const character = db
    .prepare('SELECT id FROM campaign_characters WHERE campaign_id = ? AND id = ?')
    .get(campaignId, characterId);
  if (!character) {
    return { status: 404, body: { error: 'unknown character id' } };
  }

  const itemSlug = body.item_slug;
  if (typeof itemSlug !== 'string' || itemSlug.length === 0) {
    return { status: 400, body: { error: 'item_slug must be a non-empty string' } };
  }

  const quantity = body.quantity;
  if (!isValidIntInRange(quantity, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'quantity must be a positive integer' } };
  }

  db.prepare(
    `INSERT INTO campaign_equipment (campaign_id, character_id, item_slug, quantity)
     VALUES (?, ?, ?, ?)
     ON CONFLICT (campaign_id, character_id, item_slug) DO UPDATE SET quantity = quantity + excluded.quantity`,
  ).run(campaignId, characterId, itemSlug, quantity);

  return { status: 200, body: { character_id: characterId, item_slug: itemSlug, quantity } };
}

export function inventorySummary(campaignId: string): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const partyItemsRow = db
    .prepare("SELECT COUNT(*) as count FROM campaign_inventory WHERE campaign_id = ? AND owner = 'party'")
    .get(campaignId) as { count: number };

  const assignedItemsRow = db
    .prepare('SELECT COUNT(*) as count FROM campaign_equipment WHERE campaign_id = ?')
    .get(campaignId) as { count: number };

  const partyPotionsRow = db
    .prepare(
      "SELECT COALESCE(SUM(quantity), 0) as total FROM campaign_inventory WHERE campaign_id = ? AND owner = 'party' AND item_slug = 'healing-potion'",
    )
    .get(campaignId) as { total: number };

  const assignedPotionsRow = db
    .prepare(
      "SELECT COALESCE(SUM(quantity), 0) as total FROM campaign_equipment WHERE campaign_id = ? AND item_slug = 'healing-potion'",
    )
    .get(campaignId) as { total: number };

  return {
    status: 200,
    body: {
      campaign_id: campaignId,
      party_items: partyItemsRow.count,
      assigned_items: assignedItemsRow.count,
      healing_potions_available: Math.max(0, partyPotionsRow.total - assignedPotionsRow.total),
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/inventory.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../database'
require_relative 'campaigns'

module Handlers
  # Campaign inventory and per-character equipment assignment.
  # `healing-potion` availability in the summary is the party's on-hand
  # quantity minus whatever has been assigned out to characters.
  module Inventory
    HEALING_POTION_SLUG = 'healing-potion'

    module_function

    def find_character(campaign_id, character_id)
      row = Database.query(<<~SQL).first
        SELECT id FROM campaign_characters
        WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(character_id)};
      SQL
      raise HttpError.new(404, 'unknown character id') unless row

      row
    end

    def add_inventory_item(campaign_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)

      item_slug = body['item_slug']
      quantity = body['quantity']
      owner = body['owner']

      raise HttpError.new(400, 'item_slug must be a string') unless item_slug.is_a?(String) && !item_slug.empty?
      raise HttpError.new(400, 'quantity must be a positive integer') unless quantity.is_a?(Integer) && quantity.positive?
      raise HttpError.new(400, 'owner must be a string') unless owner.is_a?(String) && !owner.empty?

      Database.exec(<<~SQL)
        INSERT INTO campaign_inventory (campaign_id, item_slug, quantity, owner)
        VALUES (
          #{Database.escape(campaign_id)},
          #{Database.escape(item_slug)},
          #{Database.int(quantity)},
          #{Database.escape(owner)}
        );
      SQL

      [201, { item_slug: item_slug, quantity: quantity, owner: owner }]
    end

    def assign_equipment(campaign_id, character_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)
      find_character(campaign_id, character_id)

      item_slug = body['item_slug']
      quantity = body['quantity']

      raise HttpError.new(400, 'item_slug must be a string') unless item_slug.is_a?(String) && !item_slug.empty?
      raise HttpError.new(400, 'quantity must be a positive integer') unless quantity.is_a?(Integer) && quantity.positive?

      Database.exec(<<~SQL)
        INSERT INTO character_equipment (campaign_id, character_id, item_slug, quantity)
        VALUES (
          #{Database.escape(campaign_id)},
          #{Database.escape(character_id)},
          #{Database.escape(item_slug)},
          #{Database.int(quantity)}
        );
      SQL

      [200, { character_id: character_id, item_slug: item_slug, quantity: quantity }]
    end

    def summary(campaign_id, _body)
      Handlers::Campaigns.find_campaign(campaign_id)

      party_items_row = Database.query(<<~SQL).first
        SELECT COUNT(*) AS count FROM campaign_inventory
        WHERE campaign_id = #{Database.escape(campaign_id)} AND owner = 'party';
      SQL
      party_items = party_items_row ? party_items_row['count'].to_i : 0

      assigned_items_row = Database.query(<<~SQL).first
        SELECT COUNT(*) AS count FROM character_equipment
        WHERE campaign_id = #{Database.escape(campaign_id)};
      SQL
      assigned_items = assigned_items_row ? assigned_items_row['count'].to_i : 0

      party_potions_row = Database.query(<<~SQL).first
        SELECT COALESCE(SUM(quantity), 0) AS total FROM campaign_inventory
        WHERE campaign_id = #{Database.escape(campaign_id)}
          AND owner = 'party'
          AND item_slug = #{Database.escape(HEALING_POTION_SLUG)};
      SQL
      party_potions = party_potions_row ? party_potions_row['total'].to_i : 0

      assigned_potions_row = Database.query(<<~SQL).first
        SELECT COALESCE(SUM(quantity), 0) AS total FROM character_equipment
        WHERE campaign_id = #{Database.escape(campaign_id)}
          AND item_slug = #{Database.escape(HEALING_POTION_SLUG)};
      SQL
      assigned_potions = assigned_potions_row ? assigned_potions_row['total'].to_i : 0

      [200, {
        campaign_id: campaign_id,
        party_items: party_items,
        assigned_items: assigned_items,
        healing_potions_available: party_potions - assigned_potions
      }]
    end
  end
end
